### PR TITLE
Feature/bmspt 181 downgrade bcf

### PR DESCRIPTION
**Pull request by DanielLepold**
# Pull Request Template

## Description

PR includes the downgrade functionality of Bcf3.0 to Bcf 2.1

## Type Of Change

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality
to not work as expected)
- [ ] This change requires a documentation update
- [ ] Code style update (formatting, local variables)
- [ ] Refactoring (no functional changes, no API changes)
- [ ] CI/Build related changes (so that Github is happy)
- [ ] Dockerfile related changes
- [ ] Documentation content changes
- [x] Tests (extended coverage or fixing existing tests)
- [ ] Updated dependencies

### API Changes

- [x] Introducing a new method or property
- [ ] Updating the API
- [ ] Deprecating methods or properties

### Environment (Docker)

- [ ] Adding a new environment variable
- [ ] Updating the production container

## Checklist

- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my own code
- [x] I have checked my code and corrected any misspellings
- [x] I have commented my code, particularly in hard-to-understand areas
- [] I have made corresponding changes to the documentation.
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally with my changes
- [ ] I have tested my code in the local docker-compose environment
- [x] My feature branch builds on Github

### CODE REVIEW

File: src/bcf-toolkit/Converter/Bcf30/SchemaConverterToBcf21.cs [hunk 1]
```diff
@@ -1,11 +1,296 @@
...
+
+    return builder.Build();
+  }
+
+  private static Model.Bcf21.ViewPoint ConvertViewPoint(
+    Model.Bcf30.ViewPoint from) {
+
+    var viewPoint = new ViewPoint();
+
+    viewPoint.VisualizationInfo =
```
Comment (balintbendebimspot) [inline src/bcf-toolkit/Converter/Bcf30/SchemaConverterToBcf21.cs]: Maybe a builder would be nice to use. There are fields that are not filled:
* Viewpoint
* Snapshot
* Index
* Guid

File: src/bcf-toolkit/Converter/Bcf30/SchemaConverterToBcf21.cs [hunk 2]
```diff
@@ -1,11 +1,296 @@
...
+      .SetPerspectiveCamera(ConvertPerspectiveCamera(from.PerspectiveCamera));
+    //TODO: SetViewSetupHints data loss
+
+    return builder.Build();
+
+  }
+
+  private static Model.Bcf21.OrthogonalCamera ConvertOrthogonalCamera(
+    Model.Bcf30.OrthogonalCamera from) {
+    var camera = new OrthogonalCamera();
```
Comment (balintbendebimspot) [inline src/bcf-toolkit/Converter/Bcf30/SchemaConverterToBcf21.cs]: Orthogonal camera builder should be used, or validation should be do manually.

File: src/bcf-toolkit/Converter/Bcf30/SchemaConverterToBcf21.cs [hunk 3]
```diff
@@ -1,11 +1,296 @@
...
+    camera.CameraViewPoint.Z = from.CameraViewPoint.Z;
+
+    camera.ViewToWorldScale = from.ViewToWorldScale;
+    //TODO: AspectRatio data loss
+    return camera;
+  }
+
+  private static Model.Bcf21.PerspectiveCamera ConvertPerspectiveCamera(
+    Model.Bcf30.PerspectiveCamera from) {
+    var camera = new PerspectiveCamera();
```
Comment (balintbendebimspot) [inline src/bcf-toolkit/Converter/Bcf30/SchemaConverterToBcf21.cs]: Perspective camera builder should be used, or do the validation manually.

File: src/bcf-toolkit/Converter/Bcf30/SchemaConverterToBcf21.cs [hunk 4]
```diff
@@ -1,11 +1,296 @@
...
+  /// </summary>
+  /// <param name="from">The object which must be converted.</param>
+  /// <returns>Returns the converted object.</returns>
+  public static Model.Bcf21.Bcf Convert(Model.Bcf30.Bcf from) {
+    var builder = new BcfBuilder();
+    builder
+      .AddMarkups(from.Markups.Select(ConvertMarkup).ToList())
+      .SetProject(ConvertProject(from.Project?.Project));
+
+    return builder.Build();
```
Comment (balintbendebimspot) [inline src/bcf-toolkit/Converter/Bcf30/SchemaConverterToBcf21.cs]: Where the builder functions are chained, the final `Build` can be called in the pipe.
```
return builder
  .AddMarkups(from.Markups.Select(ConvertMarkup).ToList())
  .SetProject(ConvertProject(from.Project?.Project));
  .Build();
```

**Review by balintbendebimspot — COMMENTED** [review]
Nice job! See my comments!

File: src/bcf-toolkit/Converter/Bcf30/SchemaConverterToBcf21.cs (hunk 4 above)
Comment (DanielLepold) [reply src/bcf-toolkit/Converter/Bcf30/SchemaConverterToBcf21.cs]: Ok. Fixed them.

File: src/bcf-toolkit/Converter/Bcf30/SchemaConverterToBcf21.cs (hunk 3 above)
Comment (DanielLepold) [reply src/bcf-toolkit/Converter/Bcf30/SchemaConverterToBcf21.cs]: Done.

File: src/bcf-toolkit/Converter/Bcf30/SchemaConverterToBcf21.cs (hunk 2 above)
Comment (DanielLepold) [reply src/bcf-toolkit/Converter/Bcf30/SchemaConverterToBcf21.cs]: Done.

File: src/bcf-toolkit/Converter/Bcf30/SchemaConverterToBcf21.cs (hunk 1 above)
Comment (DanielLepold) [reply src/bcf-toolkit/Converter/Bcf30/SchemaConverterToBcf21.cs]: Done.

**Pull request merged**